torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -729,6 +758,107 @@ void Reducer::finalize_backward() {
+  auto indices_tensor_device = at::empty({total_size + 1}, options);
+  indices_tensor_device.copy_(indices_tensor, true);
+  std::vector<at::Tensor> indices_tensor_list = {indices_tensor_device};
+  process_group_->broadcast(indices_tensor_list)->wait();
+  indices_tensor.copy_(indices_tensor_list.front());
+
+  // Update num_buckets after receiving it from rank 0
+  num_buckets = indices_accessor[indices_accessor_Index];
+
+  // Broadcast bucket_sizes
+  auto bucket_sizes_tensor = at::empty({(int64_t)num_buckets}, at::kInt);
+  auto bucket_sizes_accessor = bucket_sizes_tensor.accessor<int, 1>();
+  for (size_t i = 0; i < num_buckets; i++) {
+    // For rank != 0, it is possible that local num buckets bucket_sizes.size()
+    // is smaller than broadcasted num_buckets
+    bucket_sizes_accessor[i] =
+        bucket_sizes.at(std::min(i, (bucket_sizes.size() - 1)));
+  }
+  auto bucket_sizes_tensor_device = at::empty({(int64_t)num_buckets}, options);
+  bucket_sizes_tensor_device.copy_(bucket_sizes_tensor, true);
+  std::vector<at::Tensor> bucket_sizes_tensor_list = {
+      bucket_sizes_tensor_device};
+  process_group_->broadcast(bucket_sizes_tensor_list)->wait();
+  bucket_sizes_tensor.copy_(bucket_sizes_tensor_list.front());
+
+  // Clear bucket_indices first, and then update bucket_indices using received
+  // num_buckets, bucket_sizes_tensor and indices_tensor from rank 0
+  bucket_indices.clear();
+  bucket_indices.reserve(num_buckets);
+  indices_accessor_Index = 0;
+  for (size_t i = 0; i < num_buckets; i++) {
+    const auto& bucket_size = bucket_sizes_accessor[i];
+    std::vector<size_t> bucket;
+    bucket.reserve(bucket_size);
+    for (size_t j = 0; j < bucket_size; j++) {
+      bucket.push_back(indices_accessor[indices_accessor_Index++]);
+    }
+    bucket_indices.emplace_back(std::move(bucket));
+  }
+}
+
+std::vector<std::vector<size_t>> Reducer::rebuildBuckets() {
+  TORCH_INTERNAL_ASSERT(
+      rebuilt_params_.size() == rebuilt_param_indices_.size(),
+      "rebuilt parameter tensors size is not same as rebuilt parameter indices size.");
+  TORCH_INTERNAL_ASSERT(
+      replicas_[0].size() == rebuilt_param_indices_.size(),
+      "rebuilt parameter indices size is not same as original model parameters size.");
+  std::vector<std::vector<size_t>> rebuilt_bucket_indices;
+  std::vector<size_t> bucket_size_limits;
+  bucket_size_limits.push_back(kDefaultFirstBucketBytes);
+  bucket_size_limits.push_back(bucket_bytes_cap_);
+  rebuilt_bucket_indices = compute_bucket_assignment_by_size(
+      rebuilt_params_,
+      bucket_size_limits,
+      expect_sparse_gradients_[0],
+      rebuilt_param_indices_);
+
+  // For rebuilt bucket indices, it needs to be synced across all ranks.
+  // Broadcast the newly rebuilt bucket indices from rank 0 in default.
+  // After syncing up rebuilt bucket indices, initialize buckets for reducer.
+  sync_bucket_indices(rebuilt_bucket_indices);
+
+  has_rebuilt_bucket_ = true;
```

Comment:
nit: let's add an assert in the beginning of this function to make sure that `has_rebuilt_bucket_ ` is indeed false.